src/common/Gateway.sol

```diff
@@ -190,6 +182,18 @@ contract Gateway is Auth, IGateway, Recoverable {
         }
     }
 
+    function _handleBatch(uint16 centrifugeId, bytes memory batch_, IAdapter adapter_) internal {
+        bytes memory message = batch_;
+        for (uint256 start; start < batch_.length;) {
+            uint256 length = processor.messageLength(message);
+            message = batch_.slice(start, length);
+            start = length;
+
+            processor.handle(centrifugeId, message);
+            emit ExecuteMessage(centrifugeId, message, adapter_);
```

Comment:
Now the events won't match. `SendMessage` will include the full batch whereas `ExecuteMessage` per sub message. I would leave `ExecuteMessage` event out of the loop

Reply:
But we need it inside of the loop to distinguish between a successful message and a failed message in the incoming feature. Would you agree with the other way around? Dispatch a SendMessage per message in the batch?